### PR TITLE
Fix `DepositReserveAsset` fees payment

### PR DESCRIPTION
**Pull request by NachoPal**
The `fee` should be calculated with the reanchored asset, otherwise it could lead to a failure where the set aside fee ends up not being enough.

@acatangiu

### CODE REVIEW

**Review by liamaharon — COMMENTED** [review]
Test?

**Comment by NachoPal** [comment]
> Test?

What test would you like to see? It failing without the changes?

You can go to https://github.com/paritytech/polkadot-sdk/pull/3339 , undo the `xcm-executor` changes and run:
```
cargo test -p  asset-hub-rococo-integration-tests reserve_transfer_native_asset_from_para_to_system_para
```
It will fail due to the reason explained in this PR.

In any case, upon examining the code, it seems that the failure is potentially expected because the [beforehand calculated](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L833-L834) fee is based on  [non-reanchored assets](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L827), while the [actual calculation](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L349) is performed for the [reanchored assets](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/lib.rs#L846-L847)

**Comment by liamaharon** [comment]
`cargo test -p asset-hub-rococo-integration-tests reserve_transfer_native_asset_from_para_to_system_para` passed for me on latest master.

I see there is a test for it in a different branch, but it is preferred I think (helps with review and ensures no regression can occur) to bundle the test with the fix when merging to master.

File: polkadot/xcm/xcm-executor/src/lib.rs
Comment (acatangiu) [inline polkadot/xcm/xcm-executor/src/lib.rs:832]:
```suggestion
					self.holding.subsume_assets(to_weigh.clone());
					let to_weigh_reanchored = Self::reanchored(to_weigh, &dest, None);
					let mut message_to_weigh =
```

**Review by acatangiu — APPROVED** [review]

**Comment by acatangiu** [comment]
I agree you should add a test for this in this branch/PR, #3339 is more controversial.

Also for a regression test of this particular fix, instead of relying on a full xcm-emulator e2e test, you could add a (simpler) executor unit-test [here](https://github.com/paritytech/polkadot-sdk/blob/f3007a6d54b3d96d617a90d3624ef33f5111ebbd/polkadot/xcm/xcm-executor/integration-tests/src/lib.rs#L14).

**Comment by NachoPal** [comment]
@liamaharon @acatangiu I added a regression test. I'll merge the PR unless you request further changes.

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5298013

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 2/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5303164

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5316154

File: polkadot/xcm/xcm-executor/integration-tests/src/lib.rs
Comment (NachoPal) [reply polkadot/xcm/xcm-executor/integration-tests/src/lib.rs:334]: The previous `PriceForMessageDelivery` was flaky returning different prices.

Primarily, it was designed to return the same price for identical messages. However, reanchored messages weren't actually the same.
  - First price calculation is done for a message with the full asset amount in `ReserveAssetDeposited`
  - Actual price calculation during `send()` is done for a message where `transport_fee` has been deducted from the initial asset amount.
  
It does not explain by itself why the test was flaky. The reason for this flakiness is that the appended `SetTopic(id)` XCM instruction is calculated using the previous block as entropy. This is why any change in the runtime led to changes in the final message which might or might not generate enough transport fees to be paid.

All of this made me realise that the current method of setting aside the `transport_fee` for `DepositReserveAsset` will never be infallible for any possible `PriceForMessageDelivery` implementation. There is a circular dependency where it is impossible to know the actual `DepositReserveAsset` amount that will finally be included in the message from which the `transport_fee` is calculated. @acatangiu @franciscoaguirre 

I followed a different approach to fix the test.

Comment (acatangiu) [reply polkadot/xcm/xcm-executor/integration-tests/src/lib.rs:334]: Can you expand a bit on the transport cost variance?

The actual amount (scalar value) of the asset within the message changes/impacts the message weight? I.e. An instruction containing asset: 2 DOT is more expensive than same instruction with asset: 3 DOT? Did I understand that correctly?
Or is it an issue with `SetTopic(id)` whose transport cost is not correctly calculated during execution of `ReserveAssetDeposited`? Looking at the code I'm missing where this cost changes between when we calculate and when it is charged..

Comment (NachoPal) [reply polkadot/xcm/xcm-executor/integration-tests/src/lib.rs:334]: > An instruction containing asset: 2 DOT is more expensive than same instruction with asset: 3 DOT? Did I understand that correctly?

Not for our `PriceForMessageDelivery` implementation, because it only depends on the message size, that is why it was failing when the asset was not reanchored. 

I am just saying that potentially, a third party could make use of the `xcm-executor` with a different `PriceForMessageDelivery` implementation, and we can not guarantee that the set aside `transport_fee` approach for `DepositReserveAsset` will be infalible. It is just something to be aware of in case the design can be improved, not a big deal, not even probable, and something that shouldn't block this PR.

Comment (NachoPal) [reply polkadot/xcm/xcm-executor/integration-tests/src/lib.rs:334]: I'll proceed to merge it if the new test looks ok for you.

**Comment by Polkadot-Forum** [comment]
This pull request has been mentioned on **Polkadot Forum**. There might be relevant details there:

https://forum.polkadot.network/t/improve-xcm-development-and-release-process/6497/1

**Pull request merged**
